### DISCUSSION TITLE
Simplify and explain the installation procedure via a `docker-compose.yml`

### DISCUSSION THREAD

**Discussion by gionata** [cms-dev/cms#1245] [Ideas]
It would be useful for lazy people like me to run the CMS platform using a few commands, even without administrator's privileges.

A `docker-compose.yml` file for bootstrapping the whole system could be a great solution.
A `postgres:latest` image could be used for the database, an `ubuntu:latest` image could manage the services and document creation (LaTeX and so on), and a `ngnix:latest` image could provide the web services access. The container should provide an initial configuration including an administrator user, a test contest and a test user participating to the test contest. The order of containers and services startup must be coded in `docker-compose.yml`.
Finally, a volume could be used to manage contest resources.

The `docker-compose.yml` would provide also a valid installation procedure explanation.

**Comment by wil93** [comment]
I recently added [docker-compose.test.yml](https://github.com/cms-dev/cms/blob/master/docker-compose.test.yml) which allows to start a CMS instance and run tests against it.

It would be nice to create a similar file for the "development" use case, which would mean for example keeping the database around instead of dropping it and re-creating it at each run like we do in the testing use case. An attempt is at #1228

I also thought about providing a third file for the "production" use case, but I fear it might be too general to support natively (e.g. maybe we should support replicating the load-heavy services like CWS, load balancing them with a reverse proxy...) so I thought for now that such a use case would be better served by the contest admins writing their own `docker-compose.yml` file (or equivalent) for it.

There is also the LaTeX stuff, which isn't always needed and might cause the docker image to build more slowly / occupy more space on disk for everyone. Maybe there is some clever way to separate the "preparing the contest" from the "running the contest" and have two separate images for it?

I would like to know: what is the use case that you're looking for? Do you want to run a contest or just test CMS locally?

**Reply by gionata** [reply]
Thanks for the reply and for references the pull request #1228.

reflecting on your response, I have two use cases:

1. a container running on my laptop for preparing the contest,
2. an other running  on a local server, for managing a competition with a dozen or two participants.

I would explain better explain the use case because I think it would be of interest for my colleagues.
I am a computer science teacher who would teach using serious games (gamification of learning). There are several platform, among which the https://training.olinfo .it/, but they cannot be used directly. 
I would like to use CMS (or maybe Terry) to trigger a competition between students of the same class, who are not necessarily talented, and thus encourage them to program. CMS must be privately managed in order to:
1. propose my own problems, even very simple ones, where problems regarding computational complexity could also emerge,
2. access all submissions to manage evaluations,
3. do not expose private data of minors.

The files of the pull request will make me easier to run CMS and, in the meanwhile, I found a very useful Docker template [inside the edomora97 CMS fork](https://github.com/edomora97/cms/tree/docker/docker) .

Thanks again
